src/Faithlife.Tracing/Data/TracingDbCommand.cs

```diff
@@ -0,0 +1,170 @@
+using System;
+using System.Data;
+using System.Data.Common;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Faithlife.Tracing.Data
+{
+	internal sealed class TracingDbCommand : DbCommand
+	{
+		public TracingDbCommand(TracingDbConnection connection, DbCommand command)
+		{
+			m_connection = connection;
+			m_command = command;
+		}
+
+		protected override void Dispose(bool disposing)
+		{
+			try
+			{
+				if (disposing)
+					m_command.Dispose();
+			}
+			finally
+			{
+				base.Dispose(disposing);
+			}
+		}
+
+		public override void Cancel() => m_command.Cancel();
+
+		protected override DbParameter CreateDbParameter() => m_command.CreateParameter();
+
+		protected override DbDataReader ExecuteDbDataReader(CommandBehavior behavior) => m_command.ExecuteReader(behavior);
+
+		public override int ExecuteNonQuery()
+		{
+			var trace = BeginTrace();
+			try
+			{
+				return m_command.ExecuteNonQuery();
+			}
+			finally
+			{
+				EndTrace(trace);
+			}
+		}
+
+		private ITrace BeginTrace()
+		{
+			var sql = CommandText;
+			string rpc = null;
+			if (CommandType == CommandType.StoredProcedure || CommandType == CommandType.TableDirect)
+			{
+				rpc = sql;
+			}
+			else
+			{
+				for (int index = sql.LastIndexOf("from", StringComparison.OrdinalIgnoreCase); index > 0; index = sql.LastIndexOf("from", sql.Length - 1, StringComparison.OrdinalIgnoreCase))
```

Comment:
Will this be a problem if the executed SQL has no from?

[Example](
https://git/Logos/WebCommon/blob/cabe257322ab0c5cabe26dd3dcc56b52345c3acc/src/Logos.Common.Data.Dapper/MySqlObject.cs#L13): ```insert into {0} ({1}) values ({2}); select last_insert_id()```

Reply:
Oh yes, this is super-hacky right now.